src/mongo/platform/waitable_atomic.cpp

```diff
@@ -235,6 +238,45 @@ bool waitUntil(const void* uaddr,
     return timeoutOverflow || errno != ETIMEDOUT;
 }
 
+#elif defined(__FreeBSD__)
+
+void notifyOne(const void* uaddr) {
+    _umtx_op(const_cast<void*>(uaddr), UMTX_OP_WAKE_PRIVATE, 1, NULL, NULL);
+}
+
+void notifyMany(const void* uaddr, int nToWake) {
+    _umtx_op(const_cast<void*>(uaddr), UMTX_OP_WAKE_PRIVATE, nToWake, NULL, NULL);
+}
+
+void notifyAll(const void* uaddr) {
+    _umtx_op(const_cast<void*>(uaddr), UMTX_OP_WAKE_PRIVATE, INT_MAX, NULL, NULL);
+}
+
+bool waitUntil(const void* uaddr,
+               uint32_t old,
+               boost::optional<system_clock::time_point> deadline) {
+    struct _umtx_time umtx_deadline;
+    void* uaddr2 = nullptr;
+
+    if (deadline) {
+        umtx_deadline._timeout.tv_sec = durationCount<Seconds>(deadline->time_since_epoch());
+        umtx_deadline._timeout.tv_nsec = durationCount<Nanoseconds>(
+                deadline->time_since_epoch() - stdx::chrono::seconds(umtx_deadline._timeout.tv_sec));
+        umtx_deadline._flags = UMTX_ABSTIME;
+        umtx_deadline._clockid = CLOCK_REALTIME_FAST;
+        uaddr2 = &umtx_deadline;
+    }
+
+    int umtxOpRet;
+    if ((umtxOpRet = _umtx_op(const_cast<void*>(uaddr), UMTX_OP_WAIT_UINT_PRIVATE, old, (void*)sizeof(struct _umtx_time), uaddr2)) != 0) {
+        if (errno == ETIMEDOUT) {
```

Comment:
The `umtxOpRet` error check is incorrect. The function returns -1 on error and sets `errno`, but the code checks if `umtxOpRet != 0` which will trigger the invariant even for successful returns.